### PR TITLE
Completely remove weave from PyCBC

### PR DESCRIPTION
**Pull request by spxiwh**
This PR intends to remove completely the last weave dependencies within PyCBC. This was based on a coffee-break conversation with @josh-willis where he pointed out that you could wrap c/c++ code directly from cython. This wasn't *quite* as easy as Josh suggested, but I've been doing this as a no-internet-while-on-the-train thing in small batches over a few weeks. I've also learnt a lot about Cython, which I'm increasingly impressed with the more I learn!

At this point the code compiles without a weave dependency. I still have some questions about this, but the first cut is to check if all the automated build stuff works (or at least to fix the inevitable failures)

### CODE REVIEW

**Comment by ahnitz** [comment]
Huge props to you @spxiwh. This is a long time coming...

**Comment by spxiwh** [comment]
@josh-willis The code compiles correctly for me on AEI cluster. However, there's compilation errors here:
https://api.travis-ci.org/v3/job/587077749/log.txt
Do these errors mean anything to you? My only thought is that perhaps this is being compiled as c and not c++, but I'm not sure why that would happen.

**Comment by josh-willis** [comment]
@spxiwh Yeah, I think it is being compiled as C and not C++; you can see before all of the errors that`gcc` is being called rather than `g++`.  I don't know that I've tried building C++ explicitly via Cython, but a quick google search turned up [this page](https://cython.readthedocs.io/en/latest/src/userguide/wrapping_CPlusPlus.html) which skimming suggests we might need to do one or more of:
1. Have the first line of the `.pyx` files be:
```
# distutils: language = c++
```
1. Add `language="c++"` to the `Extension` call that builds the module in `setup.py`.
I'm going to try testing this in an hour or so but you may also want to try

**Comment by josh-willis** [comment]
@spxiwh @ahnitz I've tried adding the `# distutils: language = c++` line, as well as another `# distutils: sources = [...]` line to the top of the `.pyx` file, and I still see this error, and can see from the command line it prints that it's invoking `gcc` rather than `g++`. I also see that Ian already added the `language = 'c++'` argument to `setup.py`.  So everything I can find would tell me it should be trying to compile with C++, but it's not.  The only thing I can think of is that something about how we invoke the build is different---we do use things from `setuptools` rather than `distutils` for this purpose. But I'm not even sure what to experiment with here to try.

For the record, I'm doing this testing in a clean virtualenv on CIT (which is RHEL 7---do I remember, Ian, that the AEI cluster you're testing on would be Debian?). As part of my virtualenv, before I do anything else, I do upgrade both `pip` and `setuptools`, since I know the system versions are not current enough for our needs.

**Comment by spxiwh** [comment]
@josh-willis I think this C vs C++ think is a red herring. But I do see the problem on CIT, but not on Hypatia (Debian). This can be reduced to running the following in the top level directory of PyCBC:

```
g++ -c pycbc/events/simd_threshold_ccode.cpp -o test.o
```

which runs fine on Hypatia, but fails with the same errors as with the autobuild on CIT. The main difference I can see between the two is that CIT is using gcc v4.8.5 whereas hypatia is using v6.3.0.

However, this code used to compile fine under weave ... So I'm confused as to what the problem is

**Comment by spxiwh** [comment]
I got this to compile by adding https://github.com/gwastro/pycbc/pull/2946/commits/9db15a48d6adf9776608a30af74459d4277302d5
Not sure why this makes a difference, but apparently on a newer gcc it doesn't.

**Comment by josh-willis** [comment]
@spxiwh I'll stop for the night now, but at least 3 of 4 Travis tests are now passing, and the new codeclimate issue count has come down a little. Even once we get all of the Travis tests to pass, I'd like to test this a little more thoroughly, since it's a big change.  Probably I'll run a short workflow (over just a couple of hours of data) to see that it succeeds.

**Comment by spxiwh** [comment]
@josh-willis thanks for helping get this ready! I absolutely agree with the need to carefully test this (I'd probably want to run it through a full chunk and profile an example inspiral job). I think that I would want to cut at least another release in the `v1.14.X` series before merging this, which would then bump to `v1.15.X`.

**Comment by josh-willis** [comment]
@spxiwh I wonder if rebasing from current master will at least help with the one remaining Travis failure---it seems unrelated to changes we've been making, but rather some upstream stuff.  I've verified that I can rebase cleanly (no conflicts) but it might be better for you to do that, since the branch for this PR lives in your repo, and a force push will be required.

**Comment by spxiwh** [comment]
@josh-willis Others are also reporting that they are seeing this Travis failure. So it looks like it's something unrelated to this, which might be affecting master and that will need fixing separately.

**Comment by josh-willis** [comment]
Well with 1d73d90 all of the travis builds now succeed.  I'm not sure whether simply removing the `yum update` in `tools/docker_build_dist.sh` is the right solution.  I know that the folks who maintain the lscsoft software stack are working on upgrading that to EPEL 7.7 compatibility, including Python 3.6, but I think the target time is still a few weeks away (during the commissioning break). I'm not sure what I think the right solution is here: we are depending on a particular starting point for our image (in this case the `ligo/lalsuite-dev:el7` docker image) but then just updating all the packages in that image to the newest versions available. Should we only update what we know we need to?  But that seems like much more work to maintain.

What to @spxiwh @ahnitz @duncan-brown think?

**Comment by spxiwh** [comment]
I've just rebased this with master. I'll run one more end-to-end test now (compared against the recent v1.14.4 release). If that goes well, I would like to proceed with merging this.

**Comment by spxiwh** [comment]
@ahnitz Could you take a look over the edited top-level `__init__.py`. I think some more stuff can be removed from that. Probably easiest to make any such changes a subsequent PR.

**Comment by ahnitz** [comment]
@spxiwh Yes, there's more that can be removed from init.py. See https://github.com/gwastro/pycbc/pull/3032 which yes can be a follow-on to this PR.

File: pycbc/events/simd_threshold_ccode.cpp [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+ *                          by calling windowed_max in parallel. It then sweeps
+ *                          through the results of that (single-threaded) and
+ *                          tests for above threshold, and time-clusters the
+ *                          surviving triggers.
+ *
+ * A user calls only the last function; the other three exist to conveniently
+ * compartmentalize SIMD code from OpenMP code.
+ */
+
+#include <omp.h> // FIXME: Does this need to be optional for clang??
```
Comment (ahnitz) [inline pycbc/events/simd_threshold_ccode.cpp]: While the pragmas I believe will be ignore, I think this import yes may post a problem for clang on a default Mac setup. Unless they are now enabling omp by default?

Comment (ahnitz) [reply pycbc/events/simd_threshold_ccode.cpp]: Ok, googling says that OMP is now *enabled* in clang 3.8.0 and later. I'm not sure what Mac versions that corresponds to.

Comment (ahnitz) [reply pycbc/events/simd_threshold_ccode.cpp]: Ah, nvm, looks like Mac overrides the clang defaults and disables it...

**Comment by ahnitz** [comment]
@spxiwh cc looks fine since it's just about added FIXME's, beyond the openmp issue, if the tests pass, I think we should bite the bullet and merge.

File: pycbc/events/simd_threshold_ccode.cpp (hunk 1 above)
Comment (ahnitz) [reply pycbc/events/simd_threshold_ccode.cpp]: I think one can put this in a macro ala https://stackoverflow.com/questions/28166565/detect-gcc-as-opposed-to-msvc-clang-with-macro

So something like
```
#ifndef __clang__
    #include <omp.h>
#endif
```

Comment (spxiwh) [reply pycbc/events/simd_threshold_ccode.cpp]: I'll make sure this builds on my mac before merging!

Comment (ahnitz) [reply pycbc/events/simd_threshold_ccode.cpp]: @spxiwh You'll want to make sure you are using system clang and not from macports or elsewhere. I think those might actually have openmp. Add it to the list to stand up a MaxOSX test build at some point in travis....

**Comment by spxiwh** [comment]
@ahnitz I'm now ready to merge this, pending the MacOSX check. If you're happy please indicate approval. I have demonstrated that an end-to-end production run with/without this patch produces identical output.

File: pycbc/events/simd_threshold_ccode.cpp (hunk 1 above)
Comment (spxiwh) [reply pycbc/events/simd_threshold_ccode.cpp]: Well without these lines this doesn't build on my mac and with them it compiles, so I'm adding them and removing the FIXME. (This becomes additionally confusing as mac symlinks `gcc` to `clang` so even though it looks like you're using gcc, you're not)

**Comment by spxiwh** [comment]
Okay, mac build issue fixed, PR condensed into 3 commits (to preserve contribution from @josh-willis) all code climates (hopefully) resolved other than one issue addressed in #3032, two approvals to merge.

I'll push the button when Travis completes.

**Pull request merged**